Content.Server/Fluids/EntitySystems/PuddleSystem.Overflow.cs

```diff
@@ -0,0 +1,200 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Content.Server.Fluids.Components;
+using Content.Shared.Chemistry.Components;
+using Content.Shared.Directions;
+using Content.Shared.FixedPoint;
+using Content.Shared.Maps;
+using Content.Shared.Physics;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Map;
+using Robust.Shared.Maths;
+using Robust.Shared.Physics;
+using Robust.Shared.Utility;
+
+namespace Content.Server.Fluids.EntitySystems;
+
+public partial class PuddleSystem
+{
+    private readonly List<PuddleExpansion> _puddleExpansionQueue = new();
+    private int VisitedEntitiesLimit = 100;
+
+    struct PuddleExpansion
+    {
+        public List<PuddleComponent> PuddlesToExpand = new();
+        public readonly HashSet<Vector2i> VisitedTiles = new();
+        public Solution RemainingSolution = new();
+    }
```

Comment:
Might as well make it a class I would think?

Also we already have spreadersystem and they both do node expansions that are similar I would imagine.